sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -34,7 +34,8 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
     options.timestampFormat,
     options.zoneId,
     options.locale,
-    legacyFormat = FAST_DATE_FORMAT)
+    legacyFormat = FAST_DATE_FORMAT,
+    varLenEnabled = options.timestampFormat.contains('S'))
```

Comment:
it's only for parsing so we should always set it to true?

Reply:
I have updated this, and mv the cache optimization to `DateTimeFormatterHelper`